motr/client.c

```diff
@@ -830,6 +830,81 @@ M0_INTERNAL int m0_op_init(struct m0_op *op,
 	return M0_RC(0);
 }
 
+M0_INTERNAL int m0_calculate_md5_inc_digest(
+		struct m0_md5_inc_digest_pi *pi,
+		struct m0_pi_seed *seed,
+		struct m0_bufvec *bvec,
+		enum m0_pi_calc_flag flag,
+		unsigned char *curr_digest,
+		unsigned char *pi_value_without_seed)
+{
+	M0_ENTRY();
+
+	M0_PRE(pi != NULL);
+	M0_PRE(curr_digest != NULL);
+	M0_PRE(bvec != NULL);
+	M0_PRE(bvec->ov_vec.v_count != NULL);
+	M0_PRE(bvec->ov_buf != NULL);
+
+	/* Following code block should be removed */
+	/* sending dummy values to S3 */
+#if 1
+#define PI_DIGEST_LENGTH 92
+#define PI_VALUE_LENGTH 16
+	if (flag & M0_PI_CALC_UNIT_ZERO) {
+		 memset(&pi->pi_value, '0', PI_DIGEST_LENGTH);
+	}
+	memset(&pi->pi_value, 'A', PI_VALUE_LENGTH);
+	memset(curr_digest, '1', PI_DIGEST_LENGTH);
+	if (pi_value_without_seed) {
+		memset(pi_value_without_seed, 'B', PI_VALUE_LENGTH);
+	}
+#endif
+
+	/* Uncomment following code once motr has integration for MD5 */
+#if 0
+
+	/* This call is for first data unit, need to initialize prev_digest*/
+	if (flag & M0_PI_CALC_UNIT_ZERO) {
+		MD5_Init(&pi->prev_digest);
+	}
+
+	/* memcpy, so that we do not change the prev_digest */
+	memcpy(curr_digest, &pi->prev_digest, sizeof(MD5_CTX));
+
+	/* get the curr digest i by updating it*/
+	for (i = 0; i < bvec->ov_vec.v_nr; i++) {
```

Comment:
I feel API should take additional offset index of bufvec.
so the index i can start from offset index

If s3 has to write say 32MB data in a single write operations, then s3 will accumulate buffers each of size 16K, s3 will have to
1) Either call this api for each 16K buffers to calculate MD5 of each buffers
                                   or
2) S3 can call this API at motr unit size boundary (0MB, 1MB, 2MB... 32MB), for that s3 has to give offset index of 64K buffers ie say in first call s3 will give offset index as 0, in second call it will give as 64(64 16k buffers make 1MB), in third call 128 so on